newbank/server/LoanManager.java

```diff
@@ -0,0 +1,186 @@
+  private static final NewBankQuery_I bank = NewBank.getBank();
+
+  private HashMap<Integer, Loan> loans;
+  private int lastLoanNumber;
+
+  private LoanManager() {
+    loans = new HashMap<>();
+    lastLoanNumber = 0;
+  }
+
+  public static LoanManager getLoanManager() {
+    return loanManager;
+  }
+
+  public void clearLoans() {
+    this.loans.clear();
+    lastLoanNumber = 0;
+  }
+  
+  public String showTakenLoans(CustomerID customerID) {
+    Customer customer = bank.getCustomer(customerID);
+    String result = "";
+
+    for (Account account : customer.getAllAccounts()) {
+      for (int loanID : loans.keySet()) {
+        Loan loan = loans.get(loanID);
+        if (loan.getAccountTo() != null
+            && loan.getAccountTo().getAccountNumber().equals(account.getAccountNumber())) {
+          result = result + "Loan Number: " + loan.getNumber() + ", Account Number: "
+              + loan.getAccountFrom().getAccountNumber() + ", Amount: " + loan.getAmount()
+              + ", Interest Rate: " + loan.getInterest() + "%, Taken by: "
+              + loan.getAccountTo().getAccountNumber() + "\n";
+        }
+      }
+    }
+    return result == "" ? "No loans taken" : result;
+  }
+
+  public String paybackLoan(CustomerID customerID, int loanNumber) {
+    Loan loan = loans.get(loanNumber);
+    if (loan == null) {
+      return "Error. Invalid loan number.";
+    } else {
+      Account accountTo = loan.getAccountFrom();
+      Account accountFrom = loan.getAccountTo();
+      if (accountTo == null) {
+        return "Error. Invalid account number.";
+      }
+      if (accountFrom == null) {
+        return "Error. Invalid account number.";
+      }
+
+      return handleLoanPayment(loan, loan.getAmount());
+    }
+  }
+
+  public String paybackLoanPartial(CustomerID customerID, int loanNumber, double value) {
+    Loan loan = loans.get(loanNumber);
+    if (loan == null) {
+      return "Error. Invalid loan number.";
+    } else {
+      Account accountTo = loan.getAccountFrom();
+      Account accountFrom = loan.getAccountTo();
+      if (accountTo == null) {
+        return "Error. Invalid account number.";
+      }
+      if (accountFrom == null) {
+        return "Error. Invalid account number.";
+      }
+
+      return handleLoanPayment(loan, value);
+    }
+  }
+
+  private String handleLoanPayment(Loan loan, double value) {
+    Account accountTo = loan.getAccountFrom();
+    Account accountFrom = loan.getAccountTo();
+    double paymentAmount = value;
+    double loanAmount = loan.getAmount();
+
+    if (value <= loanAmount) {
+      loan.pay(value);
+    } else {
+      // loans.remove(loan.getNumber());
+      paymentAmount = loanAmount;
+    }
+
+    if (paymentAmount > accountFrom.getAvailableBalance()) {
+      return "Error. Insufficient funds.";
+    }
+
+    accountFrom.debit(paymentAmount);
+    accountTo.credit(paymentAmount);
+    accountTo.setFrozenAmount(paymentAmount * -1);
```

Comment:
There may be an issue here because the frozen amount was set right at the start and isn't updated when interest is added, so minusing paymentAmount will result in a negative frozen amount.

Reply:
Isn't the amount only frozen between when a loan is 'offered' and when the loan is actually taken by another customer? After the loan is taken the money is transferred to the borrower, so the frozen amount should be set to 0?

Reply:
That makes sense to me!